crucible-mir/src/Mir/Trans.hs

```diff
@@ -149,33 +149,64 @@ transConstVal _ty (Some (UsizeRepr)) (M.ConstInt i) =
-        pure c'
-    vec <- mirVector_fromVector tpr $ R.App $ E.VectorLit tpr $ V.fromList cs'
-    vecRef <- constMirRef (MirVectorRepr tpr) vec
-    ref <- subindexRef tpr vecRef (R.App $ usizeLit 0)
-    let len = R.App $ usizeLit $ fromIntegral $ length cs
-    let struct = S.mkStruct
-            (mirSliceCtxRepr tpr)
-            (Ctx.Empty Ctx.:> ref Ctx.:> len)
-    return $ MirExp (MirSliceRepr tpr) struct
-transConstVal _ty (Some (MirSliceRepr u8Repr@(C.BVRepr w))) (M.ConstStr bs)
+
+--
+-- Before the slice code got split into separate body and reference steps,
+-- it did the following after the code that's equivalent to the ConstArray
+-- case (which is where constant slice bodies get handled now):
+--
+--vecRef <- constMirRef vec_tpr vec
+--ref <- subindexRef tpr vecRef (R.App $ usizeLit 0)
+--let len = R.App $ usizeLit $ fromIntegral $ length cs
+--let struct = S.mkStruct (mirSliceCtxRepr tpr) (Ctx.Empty Ctx.:> ref Ctx.:> len)
+--return $ MirExp (MirSliceRepr tpr) struct
+--
+-- Therefore, after looking up the defid, the reference step needs to:
+--    * extract the type from the global variable it finds
+--      (note that it'll be a MirVectorRepr that it needs to unwrap)
+--    * construct a reference to the global variable
+--      (with globalMirRef rather than constMirRef, that's the point of
+--      all this)
+--    * apply subindexRef as above
+--    * cons up the length
+--    * call mkStruct
+--    * cons up the final MirExp
+--
+-- staticPlaces does the first four of these actions; addrOfPlace
+-- does the last two. Note that addrOfPlace uses mkSlice instead
+-- of mkStruct as above, but as far as I can tell it's equivalent.
+--
+transConstVal (M.TyRef (M.TySlice _ty) _) (Some (MirSliceRepr tpr)) (M.ConstSliceRef defid len) = do
+    place <- staticPlaces tpr len defid
+    addr <- addrOfPlace place
+    return addr
+
+transConstVal _ty (Some (MirVectorRepr u8Repr@(C.BVRepr w))) (M.ConstStrBody bs)
   | Just Refl <- testEquality w (knownNat @8) = do
     let bytes = map (\b -> R.App (eBVLit (knownNat @8) (toInteger b))) (BS.unpack bs)
     let vec = R.App $ E.VectorLit u8Repr (V.fromList bytes)
     mirVec <- mirVector_fromVector u8Repr vec
-    vecRef <- constMirRef (MirVectorRepr u8Repr) mirVec
-    ref <- subindexRef u8Repr vecRef (R.App $ usizeLit 0)
-    let len = R.App $ usizeLit $ fromIntegral $ BS.length bs
-    let struct = S.mkStruct
-            knownRepr
-            (Ctx.Empty Ctx.:> ref Ctx.:> len)
-    return $ MirExp (MirSliceRepr u8Repr) struct
+    let vec_tpr = MirVectorRepr u8Repr
+    return $ MirExp vec_tpr mirVec
+--
+-- This is the rest of the code that was here before the reference and
+-- body steps got split:
+--
+--vecRef <- constMirRef vec_tpr mirVec
+--ref <- subindexRef u8Repr vecRef (R.App $ usizeLit 0)
+--let len = R.App $ usizeLit $ fromIntegral $ BS.length bs
+--let struct = S.mkStruct knownRepr (Ctx.Empty Ctx.:> ref Ctx.:> len)
+--return $ MirExp (MirSliceRepr u8Repr) struct
```

Comment:
Similarly, I'd favor not including the full code from the previous implementation here. I think it suffices to say that this case it exactly like the `ConstSliceRef` above, but specialized to strings.

Reply:
This one's now gone entirely anyway.